### DISCUSSION TITLE
Feature Request: Vim support? 🙏

### DISCUSSION THREAD

**Discussion by b-n** [autometrics-dev/autometrics-shared#47] [Ideas]
Adding/having the metrics created automatically is great, however I see a huge amount of the power coming from the generation of the PromQL.

From a vim users perspective, it would be nice to be able to generate and/or open the PromQL queries directly inside vim. The way I'm used to other plugins doing this is by providing functions. E.g. rust-vim (provides functions)[https://github.com/rust-lang/rust.vim#formatting-with-rustfmt] to help formatting rust directly inline. Something that might be a little more relevant to autometrics is the `:RustPlay` function which will not only call rustplay.

In respect ot Language Servers (e.g. I know there is some thinking about that), normal ol Vim does not support LSP's out of the box. I believe NeoVim does, however some of us supplement this functionality with other plugins such as Ale.

This request comes from the Rust Amsterdam meetup on Tuesday 2nd May 2023!

**Comment by emschwartz** [comment]
Thanks @b-n for this feature request! I agree that we should have great support for autometrics in Vim.

The idea of using commands to generate the PromQL is interesting. Out of curiosity, where would you paste them to view them (Grafana, the Prometheus UI, etc)?

Related to this, @keturiosakys and @stephlow also suggested using Code Actions for the autometrics links (instead of rendering them using markdown in the tooltips). Those could either link out to Prometheus like we're doing now or have options to copy the PromQL to the clipboard.

**Reply by b-n** [reply]
I believe Code Actions are related specifically to neovim (nvim) instead of just raw old vim. That might need to be one of the first things to figure out is at what level you'd want to write a plugin. FWIW, Neovim supports LSPs out of the box, where as I am using a vim plugin called [ale](https://github.com/dense-analysis/ale) to integrate to LSPs. I'm not sure what the standard would be for building functionality on top of other plugins, and/or enhancing the output of LSPs however.

I think if you were wanting to support vim 8 OOTB, I think it's very likely you'd be going the route of travesing code in vim script, and traversing the code base to find the nearest function name, and pulling strings apart.

> The idea of using commands to generate the PromQL is interesting. Out of curiosity, where would you paste them to view them (Grafana, the Prometheus UI, etc)?

Well, maybe one step nicer. I assume with the VS Code plugin, that you're just generating a URL, from the nearest function, providing that as a link for someone to click on? If so, then `!AutoMetricsPromQL` creating that url, and just dumping it in the clipboard would work for me. Then I can past it in whatever browser that suits.

For inspiration, and from [rust-vim](https://github.com/rust-lang/rust.vim#playpen-integration), they let the vim user define the copy command via a global `g:rust_clip_command` vim variable (e.g. `let g:rust_clip_command = 'pbcopy'` as per their docs for mac)
